### PR TITLE
Fix and task

### PR DESCRIPTION
**Pull request by edikkiz**

### CODE REVIEW

File: src/components/Admin/Login.tsx [hunk 1]
```diff
@@ -32,12 +38,17 @@ const Login: FC<LoginProps> = () => {
...
         password,
       }
-      const {
-        headers: { authorization: token },
-      } = await axios.post(`/login`, payload)
-      localStorage.setItem('accessToken', token.split(' ')[1])
-      addToast('you are logged in', { appearance: 'success' })
-      history.push("/admin/orders")
+      const login = await axios.post(`/login`, payload)
+      if (login.data.role === "ADMIN") {
```
Comment (yelyzavetak) [inline src/components/Admin/Login.tsx:42]: move roles to enums

File: src/components/Admin/controllerCity/CitiesTable.tsx [hunk 2]
```diff
@@ -2,41 +2,42 @@ import React, { Component, useEffect, useState, FC } from 'react'
...
-        getCities()
-    }, [offset])
+      })
+      setCities(data)
+      setIsLoading(false)
+    }
+    getCities()
+  }, [offset])
 
   const onSubmitDelete = (deleteId: number) => {
```
Comment (yelyzavetak) [inline src/components/Admin/controllerCity/CitiesTable.tsx:37]: use useCallback

File: src/components/Admin/controllerCity/CitiesTable.tsx [hunk 3]
```diff
@@ -68,48 +69,49 @@ const CitiesTable: FC<ControllerCityTableProps> = () => {
   }
 
   const next = () => {
```
Comment (yelyzavetak) [inline src/components/Admin/controllerCity/CitiesTable.tsx:71]: use useCallback

File: src/components/Admin/controllerCity/CitiesTable.tsx [hunk 4]
```diff
@@ -68,48 +69,49 @@ const CitiesTable: FC<ControllerCityTableProps> = () => {
...
-    setOffset(offset+limit)
+    setOffset(offset + limit)
   }
 
   const after = () => {
```
Comment (yelyzavetak) [inline src/components/Admin/controllerCity/CitiesTable.tsx:75]: use useCallback

File: src/components/Admin/controllerCity/CitiesTable.tsx [hunk 5]
```diff
@@ -68,48 +69,49 @@ const CitiesTable: FC<ControllerCityTableProps> = () => {
...
+      <Preloader isLoading={isLoading} />
+      <AdminHeader />
+      <table className="wrapper_cities__table">
+        <tr>
+          <th className="table_block_id__city">id</th>
+          <th className="table_block_name__city">Name</th>
+          <Link className="link_create__city" to="/admin/navCity" title="add new city"><th className="table_link__city">+</th></Link>
+        </tr>
+        {
+          cities.map((elem) => (
```
Comment (yelyzavetak) [inline src/components/Admin/controllerCity/CitiesTable.tsx:94]: use destructor

File: src/components/Admin/controllerMaster/AllControllerMaster.tsx [hunk 6]
```diff
@@ -69,40 +70,43 @@ const AllControllerMaster: FC<ControllerMasterProps> = () => {
...
-      <div className="wrapper_controller-city">
+  return (
+    <>
+      <AdminHeader />
+    <div className="wrapper_controller-city">
       <form className="wrapper_controller-city__form" onSubmit={handleSubmit(whatControllerUse)}>
-        <input placeholder="Enter master name" className="wrapper_controller-city__form__input" {...register("name", {required: true, minLength: 3, maxLength: 30, pattern: /[A-Za-zА-Яа-я]/})} />
-          {errors?.name?.type === "required" && <p>This field is required</p>}
+        <input placeholder="Enter master name" className="wrapper_controller-city__form__input" {...register("name", { required: true, minLength: 3, maxLength: 30, pattern: /[A-Za-zА-Яа-я]/ })} />
+        {errors?.name?.type === "required" && <p>This field is required</p>}
```
Comment (yelyzavetak) [inline src/components/Admin/controllerMaster/AllControllerMaster.tsx:84]: move error types to enum and re-use

File: src/components/Admin/controllerMaster/MastersTable.tsx [hunk 7]
```diff
@@ -62,75 +63,76 @@ const MastersTable: FC<ControllerMasterTableProps> = () => {
...
-        getMasters()
+          }
+          getMasters()
         })
     }
     setIsLoading(false)
   }
-    
+
   const next = () => {
```
Comment (yelyzavetak) [inline src/components/Admin/controllerMaster/MastersTable.tsx:85]: use useCallback

File: src/components/Admin/controllerMaster/MastersTable.tsx [hunk 8]
```diff
@@ -62,75 +63,76 @@ const MastersTable: FC<ControllerMasterTableProps> = () => {
...
-    setOffset(offset+limit)
+    setOffset(offset + limit)
   }
 
   const after = () => {
```
Comment (yelyzavetak) [inline src/components/Admin/controllerMaster/MastersTable.tsx:89]: use useCallback

File: src/components/Admin/controllerMaster/MastersTable.tsx [hunk 9]
```diff
@@ -62,75 +63,76 @@ const MastersTable: FC<ControllerMasterTableProps> = () => {
...
+      <div className="wrapper_masters">
+        <table className="wrapper_masters__table">
+          <tr>
+            <th className="table_block_id__master">id</th>
+            <th className="table_block_name__master">Name</th>
+            <th className="table_block_name__master">City</th>
+            <Link className="link_create__master" to="/admin/navMaster" title="add new master"><th className="table_link">+</th></Link>
+          </tr>
+          {
+            masters.map((elem) => (
```
Comment (yelyzavetak) [inline src/components/Admin/controllerMaster/MastersTable.tsx:111]: use destructor

File: src/components/Admin/controllerOrder/AllControllerOrder.tsx [hunk 10]
```diff
@@ -119,86 +121,93 @@ const AllControllerOrder: FC<ControllerOrderProps> = () => {
...
-      <select className="wrapper_form__select" {...register("master")}>
-              {
-              putMasters.map(({ name, id, rating}) => (
-                !rating ?
+      <Preloader isLoading={isLoading} />
+
+      <form onSubmit={handleSubmit(onSubmitPut)} className="wrapper_form__form">
+        <select className="wrapper_form__select" {...register("clockSize")}>
+          {
+            clockSizes.map((elem) => (
```
Comment (yelyzavetak) [inline src/components/Admin/controllerOrder/AllControllerOrder.tsx:152]: use destructor

File: src/components/Admin/controllerOrder/OrderTable.tsx [hunk 11]
```diff
@@ -53,88 +54,104 @@ const OrderTable: FC<ControllerOrderTableProps> = () => {
...
+            <th className="table_block_name__order">start at</th>
+            <th className="table_block_name__order">end at</th>
+            <th className="table_block_name__order">price</th>
+            <th className="table_block_name__order">feedback</th>
+            <th className="table_block_name__order">rating</th>
+            <th className="table_block_name__order">status</th>
+            <th className="link_create__order">update / delete</th>
+          </tr>
+          {
+            orders.map((elem) => (
```
Comment (yelyzavetak) [inline src/components/Admin/controllerOrder/OrderTable.tsx:114]: destructor

File: src/components/Admin/controllerOrder/OrderTable.tsx [hunk 12]
```diff
@@ -53,88 +54,104 @@ const OrderTable: FC<ControllerOrderTableProps> = () => {
...
+                <th className="table_block_name__order">{`${elem.feedback === "null" ? 'no feedback' : elem.feedback}`}</th>
+                <th className="table_block_name__order">{`${elem.rating === null ? 'not rated' : elem.rating}`}</th>
+                <th className="table_block_name__order">{`${elem.status}`}</th>
+                <button type="button" onClick={() => onSubmitDelete(elem.id)} className="link_update__order"><th className="table_link__order">delete</th></button>
+                {elem.status != Status.Complited ? <>
+                  <Link to={`/admin/navOrder/${elem.id}/${elem.userId}/${elem.masterId}/${elem.cityId}/${elem.clockSizeId}/${elem.startAt}/${elem.price}`} title="update the order" className="link_update__order"><th className="table_link__order">update</th></Link>
+                  
+                </>
+                  :
+                  <>
```
Comment (yelyzavetak) [inline src/components/Admin/controllerOrder/OrderTable.tsx:134]: why do you need empty block?

File: src/components/Admin/controllerOrder/OrderTable.tsx [hunk 13]
```diff
@@ -53,88 +54,104 @@ const OrderTable: FC<ControllerOrderTableProps> = () => {
...
+        </table>
+      </div>
+      {
+        offset !== 0 ? <button className="after_button" onClick={after}>back</button> : <button className="after_button" disabled={true} onClick={after}>back</button>
+      }
+      {
+        orders.length >= limit ? <button className="next_button" onClick={next}>next</button> : <button className="next_button" disabled={true} onClick={next}>next</button>
+      }
+      {
+        orders.length === 0 ? <div>Dont have more orders</div> : <p></p>
```
Comment (yelyzavetak) [inline src/components/Admin/controllerOrder/OrderTable.tsx:151]: use &&

File: src/components/Admin/controllerUser/UsersTable.tsx [hunk 14]
```diff
@@ -2,123 +2,125 @@ import React, { Component, useEffect, useState, FC } from 'react'
...
-                }
+  const [offset, setOffset] = useState<number>(0)
+
+  const [isLoading, setIsLoading] = useState<boolean>(false)
+
+  const { addToast } = useToasts()
+
+  useEffect(() => {
+    setIsLoading(true)
+    const getUsers = async () => {
```
Comment (yelyzavetak) [inline src/components/Admin/controllerUser/UsersTable.tsx:25]: move outside the effect
use use callback for getUsers

File: src/components/Admin/controllerUser/UsersTable.tsx [hunk 15]
```diff
@@ -2,123 +2,125 @@ import React, { Component, useEffect, useState, FC } from 'react'
...
-            {
-            users.length >= limit ? <button className="next_button" onClick={next}>next</button> : <button className="next_button" disabled={true} onClick={next}>next</button>
-            }
-            {
-            users.length === 0 ? <div>Dont have more orders</div> : <p></p>
-            }
-        </div>
-    )
+      {
+        users.length === 0 ? <div>Dont have more orders</div> : <p></p>
```
Comment (yelyzavetak) [inline src/components/Admin/controllerUser/UsersTable.tsx:120]: use && to not render empty p

File: src/components/Master/masterWorkList/masterWorkList.tsx [hunk 16]
```diff
@@ -0,0 +1,137 @@
...
+  const [offset, setOffset] = useState<number>(0)
+
+  const [isLoading, setIsLoading] = useState<boolean>(false)
+
+  const { addToast } = useToasts()
+
+  useEffect(() => {
+    setIsLoading(true)
+    const getAllOrders = async () => {
+      const { data } = await axios.get<AllOrder[]>(
```
Comment (yelyzavetak) [inline src/components/Master/masterWorkList/masterWorkList.tsx:29]: the same as for getUsers

File: src/components/Master/masterWorkList/masterWorkList.tsx [hunk 17]
```diff
@@ -0,0 +1,137 @@
...
+        </table>
+      </div>
+      {
+        offset !== 0 ? <button className="after_button" onClick={after}>back</button> : <button className="after_button" disabled={true} onClick={after}>back</button>
+      }
+      {
+        orders.length >= limit ? <button className="next_button" onClick={next}>next</button> : <button className="next_button" disabled={true} onClick={next}>next</button>
+      }
+      {
+        orders.length === 0 ? <div>Dont have more orders</div> : <p></p>
```
Comment (yelyzavetak) [inline src/components/Master/masterWorkList/masterWorkList.tsx:131]: don't render empty p

File: src/components/demo/demo.tsx [hunk 18]
```diff
@@ -0,0 +1,52 @@
...
+//                 offset,
+//                 limit
+//             }
+//         }
+//         )
+//         setOrders(data)
+//     }
+
+//     getAllOrders()
+// }, [offset])
```
Comment (yelyzavetak) [inline src/components/demo/demo.tsx:33]: do you need this comments?

**Review by yelyzavetak — CHANGES_REQUESTED** [review]

File: src/components/Admin/controllerCity/CitiesTable.tsx (hunk 5 above)
Comment (edikkiz) [reply src/components/Admin/controllerCity/CitiesTable.tsx:94]: image
так?

**Review by yelyzavetak — APPROVED** [review]

**Pull request merged**